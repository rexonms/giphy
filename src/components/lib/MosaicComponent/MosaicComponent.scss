.container {
  .gifs {
    @media (min-width:320px)  { column-count: 1; /* smartphones, iPhone, portrait 480x320 phones */ }
    @media (min-width:481px)  { column-count: 2;/* portrait e-readers (Nook/Kindle), smaller tablets @ 600 or @ 640 wide. */ }
    @media (min-width:641px)  { column-count: 3;/* portrait tablets, portrait iPad, landscape e-readers, landscape 800x480 or 854x480 phones */ }
    @media (min-width:961px)  { column-count: 4;/* tablet, landscape iPad, lo-res laptops ands desktops */ }
    column-count: 5;
    column-gap: 1px;
    div{
      width:  100%;
      display: inline-block;
      img {
        width: 100%;
      }
    }
  }
}
