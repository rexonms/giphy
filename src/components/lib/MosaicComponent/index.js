import React, { Component } from 'react';
import PropTypes from 'prop-types';

import Styles from './MosaicComponent.scss';
import GifWithInfo from '../GifWithInfo';
import LoadingBars from '../LoadingBars';

const propTypes = {
  /** Array of gifs to display */
  gifs: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  /** Tells what type of query is currently visible in the mosasic */
  endpoint: PropTypes.shape({
    name: PropTypes.string.isRequired,
    query: PropTypes.string,
  }).isRequired,
  /** Action that needs to be triggered for lazy loading */
  action: PropTypes.func.isRequired,
};

const defaultProps = {};

class MosaicComponent extends Component {
  constructor() {
    super();
    this.state = {
      fetching: false,
    };
    this.handleOnScroll = this.handleOnScroll.bind(this);
  }
  componentDidMount() {
    window.addEventListener('scroll', this.handleOnScroll);
  }
  componentWillUnmount() {
    window.removeEventListener('scroll', this.handleOnScroll);
  }
  componentWillReceiveProps() {
    // Making sure the action is called only every second
    setTimeout(() => {
      this.setState({ fetching: false });
    }, 1000);
  }
  handleOnScroll() {
    // http://stackoverflow.com/questions/9439725/javascript-how-to-detect-if-browser-window-is-scrolled-to-bottom
    const docEl = document.documentElement;
    const scrollTop = (docEl && docEl.scrollTop) || document.body.scrollTop;
    const scrollHeight = (docEl && docEl.scrollHeight) || document.body.scrollHeight;
    const clientHeight = docEl.clientHeight || window.innerHeight;
    const scrolledToBottom = Math.ceil(scrollTop + clientHeight) >= scrollHeight;
    if (scrolledToBottom && !this.state.fetching) {
      this.props.action(this.props.endpoint);
      this.setState({ fetching: true });
    }
  }
  render() {
    return (
      <div className={`${Styles.container}`}>
        <div className={Styles.gifs}>
          {this.props.gifs.map(gif => <GifWithInfo key={gif.id} {...gif} />)}
        </div>
        {this.state.fetching &&
          <LoadingBars />
        }
      </div>
    );
  }
}

MosaicComponent.propTypes = propTypes;
MosaicComponent.defaultProps = defaultProps;

export default MosaicComponent;
