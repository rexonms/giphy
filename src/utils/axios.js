import axios from 'axios';

export const foo = 'bazz';
export const getDataFromGiphy = async (url) => {
  const response = await axios.get(url);
  return response;
};

export const giphyDefaults = {
  baseURL: process.env.REACT_APP_GIPHY_BASE_URL,
  apiKey: process.env.REACT_APP_GIPHY_API_KEY,
  limit: 25,
  rating: 'g',
  offset: 0,
  endpoints: {
    trending: 'trending',
    search: 'search',
  },
};
