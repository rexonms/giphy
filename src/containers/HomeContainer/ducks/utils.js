import { giphyDefaults } from '../../../utils/axios';

export const getUrl = ({ home, endpoint }) => {
  const base = `${giphyDefaults.baseURL}/v1/gifs/${endpoint.name}`;
  const api = `api_key=${giphyDefaults.apiKey}`;
  const rating = `rating=${giphyDefaults.rating}`;
  const oldEndpointName = home.giphy && home.giphy.endpoint.name;
  let isNewEndpoint = false;
  if (oldEndpointName === undefined) {
    isNewEndpoint = true;
  }
  console.log('isNewEndpoint', isNewEndpoint);
  const offsetValue = isNewEndpoint === true ? 0 : home.giphy.pagination.offset + 1;
  const offset = `offset=${offsetValue}`;
  const url = `${base}?${api}&${rating}`;
  let limit;
  if (endpoint.name === giphyDefaults.endpoints.trending) {
    // check if there is any
    if (!home.giphy) {
      limit = `limit=${giphyDefaults.limit}`;
    } else {
      limit = `limit=${home.giphy.pagination.count + giphyDefaults.limit}`;
    }
    return `${url}&${limit}`;
  }
  console.log('offset', offset);
  console.log('endpoint', endpoint);
  return `${url}&q=${endpoint.query}`;
};

export const extractResponseInfo = (data, endpoint) => (
  {
    endpoint,
    pagination: data.pagination,
    data: data.data.map(d => (
      {
        id: d.id,
        title: d.title,
        rating: d.rating,
        user: {
          avatarUrl: d.user ? d.user.avatar_url : 'null',
        },
        image: {
          url: d.images.fixed_height_downsampled,
        },
      }
    )),
  }
);
