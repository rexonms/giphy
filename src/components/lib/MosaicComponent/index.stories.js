import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs } from '@storybook/addon-knobs/react';
import { withInfo } from '@storybook/addon-info';
import { action } from '@storybook/addon-actions';

import MosaicComponent from './index';
import { giphyDefaults } from '../../../utils/axios';

const stories = storiesOf('lib/MosaicComponent', module);
const props = {
  endpoint: {
    name: giphyDefaults.endpoints.trending,
    query: 'foo',
  },
  gifs: [
    {
      id: 'l4FGqDtRMMjOKdFGU',
      title: 'tina fey shades GIF by Unbreakable Kimmy Schmidt',
      rating: 'g',
      user: {
        avatarUrl: 'https://media.giphy.com/avatars/kimmyschmidt/8Wnie77cnpH5.png',
      },
      image: {
        url: {
          url: 'https://media2.giphy.com/media/l4FGqDtRMMjOKdFGU/200_d.gif',
          width: '356',
          height: '200',
          size: '256421',
          webp: 'https://media2.giphy.com/media/l4FGqDtRMMjOKdFGU/200_d.webp',
          webp_size: '66136',
        },
      },
    },
    {
      id: 'yoJC2B1sHdXJjPTnEs',
      title: 'leonardo dicaprio omg GIF by WE tv',
      rating: 'g',
      user: {
        avatarUrl: 'https://media.giphy.com/avatars/wetv/ez6Knwamyn0A.jpg',
      },
      image: {
        url: {
          url: 'https://media3.giphy.com/media/yoJC2B1sHdXJjPTnEs/200_d.gif',
          width: '259',
          height: '200',
          size: '247691',
          webp: 'https://media3.giphy.com/media/yoJC2B1sHdXJjPTnEs/200_d.webp',
          webp_size: '58832',
        },
      },
    },
    {
      id: 'b1YFCoOD2aX2E',
      title: 'frances mcdormand smiling GIF by Maudit',
      rating: 'g',
      user: {
        avatarUrl: 'https://media0.giphy.com/avatars/maudit.gif',
      },
      image: {
        url: {
          url: 'https://media2.giphy.com/media/b1YFCoOD2aX2E/200_d.gif',
          width: '345',
          height: '200',
          size: '353795',
          webp: 'https://media2.giphy.com/media/b1YFCoOD2aX2E/200_d.webp',
          webp_size: '102774',
        },
      },
    },
  ],
  action: () => {},
};

stories.addDecorator(withKnobs);
stories.addDecorator(story => (story()));
stories.add('MosaicComponent', withInfo('Component')(() => (
  <MosaicComponent
    {...props}
    action={action('action')}
  />
)));
