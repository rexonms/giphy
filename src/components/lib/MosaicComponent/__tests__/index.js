import React from 'react';
import { shallow } from 'enzyme';

import MosaicComponent from './../index';
import { giphyDefaults } from '../../../../utils/axios';

describe('Welcome Component', () => {
  let props;
  let shallowMosaicContainer;
  const mosaicContainer = () => {
    if (!shallowMosaicContainer) {
      shallowMosaicContainer = shallow(<MosaicComponent {...props} />);
    }
    return shallowMosaicContainer;
  };
  beforeEach(() => {
    props = {
      endpoint: {
        name: giphyDefaults.endpoints.trending,
        query: 'foo',
      },
      gifs: [
        {
          id: 'l4FGqDtRMMjOKdFGU',
          title: 'tina fey shades GIF by Unbreakable Kimmy Schmidt',
          rating: 'g',
          user: {
            avatarUrl: 'https://media.giphy.com/avatars/kimmyschmidt/8Wnie77cnpH5.png',
          },
          image: {
            url: {
              url: 'https://media2.giphy.com/media/l4FGqDtRMMjOKdFGU/200_d.gif',
              width: '356',
              height: '200',
              size: '256421',
              webp: 'https://media2.giphy.com/media/l4FGqDtRMMjOKdFGU/200_d.webp',
              webp_size: '66136',
            },
          },
        },
        {
          id: 'yoJC2B1sHdXJjPTnEs',
          title: 'leonardo dicaprio omg GIF by WE tv',
          rating: 'g',
          user: {
            avatarUrl: 'https://media.giphy.com/avatars/wetv/ez6Knwamyn0A.jpg',
          },
          image: {
            url: {
              url: 'https://media3.giphy.com/media/yoJC2B1sHdXJjPTnEs/200_d.gif',
              width: '259',
              height: '200',
              size: '247691',
              webp: 'https://media3.giphy.com/media/yoJC2B1sHdXJjPTnEs/200_d.webp',
              webp_size: '58832',
            },
          },
        },
        {
          id: 'b1YFCoOD2aX2E',
          title: 'frances mcdormand smiling GIF by Maudit',
          rating: 'g',
          user: {
            avatarUrl: 'https://media0.giphy.com/avatars/maudit.gif',
          },
          image: {
            url: {
              url: 'https://media2.giphy.com/media/b1YFCoOD2aX2E/200_d.gif',
              width: '345',
              height: '200',
              size: '353795',
              webp: 'https://media2.giphy.com/media/b1YFCoOD2aX2E/200_d.webp',
              webp_size: '102774',
            },
          },
        },
      ],
      action: jest.fn(),
    };
  });

  it('should render previous snapshot', () => {
    expect(mosaicContainer()).toMatchSnapshot();
  });
  it('should render with proper className', () => {
    expect(mosaicContainer().find('.container').exists()).toBe(true);
  });
  it('should have child components', () => {
    expect(mosaicContainer().find('GifWithInfo').exists()).toBe(true);
    expect(mosaicContainer().find('LoadingBars').exists()).not.toBe(true);
  });
});

