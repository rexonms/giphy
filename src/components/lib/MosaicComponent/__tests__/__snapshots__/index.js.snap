// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome Component should render previous snapshot 1`] = `
<div
  className="container"
>
  <div
    className="gifs"
  >
    <GifWithInfo
      id="l4FGqDtRMMjOKdFGU"
      image={
        Object {
          "url": Object {
            "height": "200",
            "size": "256421",
            "url": "https://media2.giphy.com/media/l4FGqDtRMMjOKdFGU/200_d.gif",
            "webp": "https://media2.giphy.com/media/l4FGqDtRMMjOKdFGU/200_d.webp",
            "webp_size": "66136",
            "width": "356",
          },
        }
      }
      key="l4FGqDtRMMjOKdFGU"
      rating="g"
      title="tina fey shades GIF by Unbreakable Kimmy Schmidt"
      user={
        Object {
          "avatarUrl": "https://media.giphy.com/avatars/kimmyschmidt/8Wnie77cnpH5.png",
        }
      }
    />
    <GifWithInfo
      id="yoJC2B1sHdXJjPTnEs"
      image={
        Object {
          "url": Object {
            "height": "200",
            "size": "247691",
            "url": "https://media3.giphy.com/media/yoJC2B1sHdXJjPTnEs/200_d.gif",
            "webp": "https://media3.giphy.com/media/yoJC2B1sHdXJjPTnEs/200_d.webp",
            "webp_size": "58832",
            "width": "259",
          },
        }
      }
      key="yoJC2B1sHdXJjPTnEs"
      rating="g"
      title="leonardo dicaprio omg GIF by WE tv"
      user={
        Object {
          "avatarUrl": "https://media.giphy.com/avatars/wetv/ez6Knwamyn0A.jpg",
        }
      }
    />
    <GifWithInfo
      id="b1YFCoOD2aX2E"
      image={
        Object {
          "url": Object {
            "height": "200",
            "size": "353795",
            "url": "https://media2.giphy.com/media/b1YFCoOD2aX2E/200_d.gif",
            "webp": "https://media2.giphy.com/media/b1YFCoOD2aX2E/200_d.webp",
            "webp_size": "102774",
            "width": "345",
          },
        }
      }
      key="b1YFCoOD2aX2E"
      rating="g"
      title="frances mcdormand smiling GIF by Maudit"
      user={
        Object {
          "avatarUrl": "https://media0.giphy.com/avatars/maudit.gif",
        }
      }
    />
  </div>
</div>
`;
